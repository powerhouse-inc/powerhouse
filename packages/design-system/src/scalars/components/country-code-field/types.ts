export interface CountryCodeProps {
  onChange?: (value: string | string[]) => void;
  placeholder?: string;
  allowedCountries?: string[];
  excludedCountries?: string[];
  includeDependentAreas?: boolean;
  viewMode?: "CodesOnly" | "NamesOnly" | "NamesAndCodes";
  showFlagIcons?: boolean;
  enableSearch?: boolean;
}
