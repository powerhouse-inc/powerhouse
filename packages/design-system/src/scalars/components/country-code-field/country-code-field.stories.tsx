import type { Meta, StoryObj } from "@storybook/react";
import { withForm } from "../../lib/decorators.js";
import {
  getDefaultArgTypes,
  getValidationArgTypes,
  PrebuiltArgTypes,
  StorybookControlCategory,
} from "../../lib/storybook-arg-types.js";
import { CountryCodeField } from "./country-code-field.js";

const meta: Meta<typeof CountryCodeField> = {
  title: "Document Engineering/Scalars/Country Code Field",
  component: CountryCodeField,
  decorators: [
    withForm,
    (Story) => (
      <div style={{ maxWidth: "280px", margin: "1rem auto 0" }}>
        <Story />
      </div>
    ),
  ],
  parameters: {
    layout: "padded",
  },
  tags: ["autodocs"],
  argTypes: {
    ...getDefaultArgTypes(),
    ...PrebuiltArgTypes.placeholder,

    allowedCountries: {
      control: "object",
      description: "List of allowed country codes",
      table: {
        type: { summary: "string[]" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    excludedCountries: {
      control: "object",
      description: "List of country codes to exclude",
      table: {
        type: { summary: "string[]" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    includeDependentAreas: {
      control: "boolean",
      description: "Whether to include dependent territories in country list",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "false" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    viewMode: {
      control: "radio",
      options: ["CodesOnly", "NamesOnly", "NamesAndCodes"],
      description: "How to display country options in dropdown",
      table: {
        type: { summary: '"CodesOnly" | "NamesOnly" | "NamesAndCodes"' },
        defaultValue: { summary: '"NamesOnly"' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    showFlagIcons: {
      control: "boolean",
      description: "Whether to show country flag icons",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "true" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },
    enableSearch: {
      control: "boolean",
      description: "Enable search functionality",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "false" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    ...getValidationArgTypes(),
  },
  args: {
    name: "country-code-field",
    errors: [],
    warnings: [],
  },
} satisfies Meta<typeof CountryCodeField>;

export default meta;

type Story = StoryObj<typeof CountryCodeField>;

export const Default: Story = {
  args: {
    label: "Country",
    placeholder: "Select a country",
  },
};

export const Hovered: Story = {
  args: {
    label: "Country",
    placeholder: "Select a country",
  },
  parameters: {
    pseudo: { hover: true },
  },
};

export const Focused: Story = {
  args: {
    label: "Country",
    placeholder: "Select a country",
  },
  parameters: {
    pseudo: { focus: true },
  },
};

export const Disabled: Story = {
  args: {
    label: "Disabled field",
    value: "FR",
    disabled: true,
  },
};

export const Required: Story = {
  args: {
    label: "Country",
    placeholder: "Select a country",
    required: true,
  },
};

export const WithSearchEnabled: Story = {
  args: {
    label: "Country",
    description: "Type to search through options",
    placeholder: "Select a country",
    enableSearch: true,
  },
};

export const WithError: Story = {
  args: {
    label: "Country",
    value: "AO",
    errors: ["Please select a different country"],
  },
};

export const WithWarning: Story = {
  args: {
    label: "Country",
    value: "AF",
    warnings: ["This country may have restricted access"],
  },
};

export const WithoutFlags: Story = {
  args: {
    label: "Country",
    description: "Shows country options without flag icons",
    placeholder: "Select a country",
    showFlagIcons: false,
  },
};

export const WithCodesOnly: Story = {
  args: {
    label: "Country",
    description: "Shows country codes only",
    placeholder: "Select a country",
    viewMode: "CodesOnly",
  },
};

export const WithNamesAndCodes: Story = {
  args: {
    label: "Country",
    description: "Shows country names and codes",
    placeholder: "Select a country",
    viewMode: "NamesAndCodes",
  },
};

export const WithDependentAreas: Story = {
  args: {
    label: "Country",
    description: "Shows dependent areas",
    placeholder: "Select a country",
    includeDependentAreas: true,
  },
};
