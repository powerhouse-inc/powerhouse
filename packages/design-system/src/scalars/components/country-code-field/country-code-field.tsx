import { SelectFieldRaw } from "#scalars";
import React from "react";
import { CircleFlag } from "react-circle-flags";
import countries, { type Countries } from "world-countries";
import { withFieldValidation } from "../fragments/with-field-validation/with-field-validation.js";
import type { FieldErrorHandling, InputBaseProps } from "../types.js";
import type { CountryCodeProps } from "./types.js";

type CountryCodeFieldBaseProps = Omit<
  React.ButtonHTMLAttributes<HTMLButtonElement>,
  | keyof InputBaseProps<string>
  | keyof FieldErrorHandling
  | keyof CountryCodeProps
>;

export type CountryCodeFieldProps = CountryCodeFieldBaseProps &
  InputBaseProps<string> &
  FieldErrorHandling &
  CountryCodeProps;

const CountryCodeFieldRaw = React.forwardRef<
  HTMLButtonElement,
  CountryCodeFieldProps
>(
  (
    {
      onChange,
      placeholder,
      allowedCountries,
      excludedCountries,
      includeDependentAreas = false,
      viewMode = "NamesOnly",
      showFlagIcons = true,
      enableSearch,
      ...props
    },
    ref,
  ) => {
    const defaultOptions = (countries as unknown as Countries)
      .filter(
        (country) =>
          (includeDependentAreas ? true : country.independent) &&
          country.cca2 !== "AQ", // exclude Antarctica
      )
      .map((country) => ({
        value: country.cca2,
        label:
          viewMode === "CodesOnly"
            ? country.cca2
            : viewMode === "NamesAndCodes"
              ? `${country.name.common} (${country.cca2})`
              : country.name.common,
        icon: showFlagIcons
          ? () => (
              <CircleFlag
                className="size-4"
                countryCode={country.cca2.toLowerCase()}
                height={16}
              />
            )
          : undefined,
      }))
      .sort((a, b) => (a.label > b.label ? 1 : a.label < b.label ? -1 : 0));

    const options =
      Array.isArray(allowedCountries) || Array.isArray(excludedCountries)
        ? defaultOptions.filter(
            (option) =>
              (!allowedCountries || allowedCountries.includes(option.value)) &&
              !excludedCountries?.includes(option.value),
          )
        : defaultOptions;

    return (
      <SelectFieldRaw
        ref={ref}
        options={options}
        selectionIcon="checkmark"
        selectionIconPosition="right"
        searchable={enableSearch}
        onChange={onChange}
        placeholder={placeholder}
        {...props}
      />
    );
  },
);

export const CountryCodeField = withFieldValidation<CountryCodeFieldProps>(
  CountryCodeFieldRaw,
  {
    validations: {
      _validOption:
        ({ allowedCountries, excludedCountries, includeDependentAreas }) =>
        (value: string | undefined) => {
          if (value === "" || value === undefined) {
            return true;
          }

          const validCountries = (countries as unknown as Countries)
            .filter(
              (country) =>
                (includeDependentAreas ? true : country.independent) &&
                country.cca2 !== "AQ",
            )
            .map((country) => country.cca2);

          // First check if it's a valid country code
          if (!validCountries.includes(value)) {
            return "Please select a valid country";
          }
          // Check if country is in allowed list
          if (
            Array.isArray(allowedCountries) &&
            !allowedCountries.includes(value)
          ) {
            return "Please select a valid country";
          }
          // Check if country is in excluded list
          if (
            Array.isArray(excludedCountries) &&
            excludedCountries.includes(value)
          ) {
            return "Please select a valid country";
          }

          return true;
        },
    },
  },
);

CountryCodeField.displayName = "CountryCodeField";
