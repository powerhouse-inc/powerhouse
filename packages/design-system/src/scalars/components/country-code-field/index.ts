export * from "./country-code-field.js";
