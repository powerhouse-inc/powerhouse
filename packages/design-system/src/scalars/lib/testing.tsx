import { render, type RenderResult } from "@testing-library/react";
import { Form } from "../components/index.js";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
