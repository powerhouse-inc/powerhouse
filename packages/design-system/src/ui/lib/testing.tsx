import { Form } from "@powerhousedao/design-system";
import type { RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
