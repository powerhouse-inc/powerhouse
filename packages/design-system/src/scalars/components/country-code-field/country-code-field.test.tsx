import { screen, waitFor } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";
import { renderWithForm } from "../../lib/testing.js";
import { CountryCodeField } from "./country-code-field.js";

describe("CountryCodeField Component", () => {
  const defaultProps = {
    name: "country",
    label: "Select Country",
  };
  window.HTMLElement.prototype.scrollIntoView = vi.fn();
  window.Element.prototype.scrollTo = vi.fn();

  it("should match snapshot", () => {
    const { asFragment } = renderWithForm(
      <CountryCodeField {...defaultProps} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with label", () => {
    renderWithForm(<CountryCodeField {...defaultProps} />);
    expect(screen.getByText("Select Country")).toBeInTheDocument();
  });

  it("should render with description", () => {
    renderWithForm(
      <CountryCodeField
        {...defaultProps}
        description="Please select your country"
      />,
    );
    expect(screen.getByText("Please select your country")).toBeInTheDocument();
  });

  it("should show required indicator when required prop is true", () => {
    renderWithForm(<CountryCodeField {...defaultProps} required />);
    expect(screen.getByText("*")).toBeInTheDocument();
  });

  it("should disable the field when disabled prop is true", () => {
    renderWithForm(<CountryCodeField {...defaultProps} disabled />);
    expect(screen.getByRole("combobox")).toBeDisabled();
  });

  it("should display error messages", async () => {
    renderWithForm(
      <CountryCodeField {...defaultProps} errors={["Country is required"]} />,
    );
    await waitFor(() =>
      expect(screen.getByText("Country is required")).toBeInTheDocument(),
    );
  });

  it("should display warning messages", () => {
    renderWithForm(
      <CountryCodeField
        {...defaultProps}
        warnings={["Some countries may not be available"]}
      />,
    );
    expect(
      screen.getByText("Some countries may not be available"),
    ).toBeInTheDocument();
  });

  it("should handle value changes", async () => {
    const user = userEvent.setup();
    const onChange = vi.fn();
    renderWithForm(<CountryCodeField {...defaultProps} onChange={onChange} />);

    const select = screen.getByRole("combobox");
    await user.click(select);
    await user.click(screen.getByText("United States"));
    expect(onChange).toHaveBeenCalledWith("US");
  });

  it("should filter countries based on search when enableSearch is true", async () => {
    const user = userEvent.setup();
    renderWithForm(
      <CountryCodeField
        {...defaultProps}
        enableSearch
        placeholder="Search..."
      />,
    );

    await user.click(screen.getByRole("combobox"));
    await user.type(screen.getByPlaceholderText("Search..."), "united");

    expect(screen.getByText("United States")).toBeInTheDocument();
    expect(screen.getByText("United Kingdom")).toBeInTheDocument();
    expect(screen.queryByText("France")).not.toBeInTheDocument();
    expect(screen.queryByText("Germany")).not.toBeInTheDocument();
  });

  it("should filter countries based on allowedCountries prop", async () => {
    const user = userEvent.setup();
    renderWithForm(
      <CountryCodeField {...defaultProps} allowedCountries={["US", "GB"]} />,
    );

    const select = screen.getByRole("combobox");
    await user.click(select);

    expect(screen.getByText("United States")).toBeInTheDocument();
    expect(screen.getByText("United Kingdom")).toBeInTheDocument();
    expect(screen.queryByText("France")).not.toBeInTheDocument();
  });

  it("should filter out countries based on excludedCountries prop", async () => {
    const user = userEvent.setup();
    renderWithForm(
      <CountryCodeField {...defaultProps} excludedCountries={["FR", "DE"]} />,
    );

    const select = screen.getByRole("combobox");
    await user.click(select);

    expect(screen.getByText("United States")).toBeInTheDocument();
    expect(screen.getByText("United Kingdom")).toBeInTheDocument();
    expect(screen.queryByText("France")).not.toBeInTheDocument();
    expect(screen.queryByText("Germany")).not.toBeInTheDocument();
  });

  it("should handle both allowedCountries and excludedCountries props", async () => {
    const user = userEvent.setup();
    renderWithForm(
      <CountryCodeField
        {...defaultProps}
        allowedCountries={["US", "GB", "FR"]}
        excludedCountries={["FR"]}
      />,
    );

    const select = screen.getByRole("combobox");
    await user.click(select);

    expect(screen.getByText("United States")).toBeInTheDocument();
    expect(screen.getByText("United Kingdom")).toBeInTheDocument();
    expect(screen.queryByText("France")).not.toBeInTheDocument();
  });

  it("should include dependent areas when includeDependentAreas is true", async () => {
    const user = userEvent.setup();
    const onChange = vi.fn();

    renderWithForm(
      <CountryCodeField
        {...defaultProps}
        includeDependentAreas
        onChange={onChange}
      />,
    );

    const select = screen.getByRole("combobox");
    await user.click(select);

    expect(screen.getByText("Puerto Rico")).toBeInTheDocument(); // US territory
    expect(screen.getByText("Guam")).toBeInTheDocument(); // US territory

    await user.click(screen.getByText("Puerto Rico"));
    expect(onChange).toHaveBeenCalledWith("PR");

    expect(screen.getByText("Puerto Rico")).toBeInTheDocument();
  });
});
