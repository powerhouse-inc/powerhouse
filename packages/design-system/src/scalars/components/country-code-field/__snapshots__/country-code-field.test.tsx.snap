// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CountryCodeField Component > should match snapshot 1`] = `
<DocumentFragment>
  <form
    id=":r0:"
    novalidate=""
  >
    <div
      class="flex flex-col gap-2"
    >
      <label
        class="inline-flex items-center text-sm font-medium leading-4 text-gray-900 dark:text-gray-50 mb-[3px]"
        for=":r1:-select"
      >
        Select Country
      </label>
      <button
        aria-controls="radix-:r2:"
        aria-expanded="false"
        aria-haspopup="dialog"
        aria-invalid="false"
        class="gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none disabled:pointer-events-none [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 flex h-9 w-full items-center justify-between px-3 py-2 dark:border-charcoal-700 dark:bg-charcoal-900 rounded-md border border-gray-300 bg-white hover:border-gray-300 hover:bg-gray-100 dark:hover:border-charcoal-700 dark:hover:bg-charcoal-800 dark:focus:ring-charcoal-300 focus:outline-none focus:ring-1 focus:ring-gray-900 focus:ring-offset-0 dark:focus-visible:ring-charcoal-300 focus-visible:ring-1 focus-visible:ring-gray-900 focus-visible:ring-offset-0"
        data-state="closed"
        id=":r1:-select"
        name="country"
        role="combobox"
        type="button"
      >
        <div
          class="mx-auto flex w-full items-center justify-end"
        >
          <svg
            class="cursor-pointer text-gray-700 dark:text-gray-400"
            fill="none"
            stroke="currentcolor"
            style="width: 16px; height: 16px;"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 9L12 15L18 9"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </button>
    </div>
  </form>
</DocumentFragment>
`;
